### PR TITLE
[JUJU-4126] Add UpdateExternalControllerModel and replace the usages of SaveAndMoveModels on facades

### PR DESCRIPTION
**Pull request by nvinuesa**
This patch adds a new domain method for external controllers: UpdateExternalControllerModel, which updates the external_model table with the model UUID and the controller UUID it is linked to.

This is a first step needed to completely remove the legacy mongodb state for external controllers.

## Checklist

- [X] Code style: imports ordered, good names, simple structure, etc
- [X] Comments saying why design decisions were made
- [X] Go unit tests, with comments saying what you're testing
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

First, let's create two controllers (one with influxdb, other with grafana) and integrate them:
```
juju bootstrap localhost ctrl-src1
juju add-model model-src1
juju deploy influxdb
juju offer influxdb:grafana-source  
---
juju bootstrap localhost ctrl-src2
juju add-model model-src2
juju deploy grafana
juju integrate grafana ctrl-src1:admin/model-src1.influxdb
```
Now you can check the status with `--relations` to validate that the integration worked.

Also, let's check that the external controller have been stored in dqlite and not in mongodb:
(still on ctrl-src2)
connect to controller machine and run dqlite:
```
lxc exec juju-xxx bash
apt install go-dqlite
dqlite -s 127.0.0.1:17666 controller
SELECT * FROM external_controller; // Here you should see one external controller:
dqlite> SELECT * FROM external_controller;
b350dd86-cff7-4e42-8979-f8b40d5907e0|ctrl-src1|-----BEGIN CERTIFICATE-----
MIIEEzCCAnugAwIBAgIVALivOnMNMuM2SgZNOFE1dgDwKoZ3MA0GCSqGSIb3DQEB
CwUAMCExDTALBgNVBAoTBEp1anUxEDAOBgNVBAMTB2p1anUtY2EwHhcNMjMwNzI0
MDcyNzU3WhcNMzMwNzI0MDczMjU3WjAhMQ0wCwYDVQQKEwRKdWp1MRAwDgYDVQQD
EwdqdWp1LWNhMIIBojANBgkqhkiG9w0BAQEFAAOCAY8AMIIBigKCAYEA0/uqd0Ac
aINPKp/lhdCKDAfBFKK5HKIt0GmSXBWa0xHwOwKi7a/h3qUhwAG4xJnLGGZc0WUY
mT8ciaJhM9uH41oopkLOvZnI8GU+D+srb1uyvYHIp9v/yICy/K61glUyoiu4ckf/
KeiPPXEZbd3u2Jz9roZPTvPgfes9Otuft3+82Dx6oW8vA4WZb/CGnRZ/tGr9oQ4M
Qqe+TbNM/kxeafyQ2c0yjVG2v3fdaAVT7PE0d8uiWQxVvcQGAyPHr2RPhhqxILJ+
YXTwKo+v46TlJj33GBKUQJfA2oGbtyv9Y7ybkhWfFsWIEmmckZU0Km97N9aGtnqX
NPF1Q85WG3OmO+odiYQ/nDpC/kL9ohndEA1HK7rzOLIk/VynHXEgzMK1Oi5U9aJH
rZ6eecyqKJ5QOkvsBCiA3zzsqRnIr/xAxnsoVTHmlIHaWU/b0hA/l+9jin+1OpbC
H3e14faaDewcsejBP5VdQjTpM+fo8Hni1lnZutxBB2PXt6yNlYt7HqaPAgMBAAGj
QjBAMA4GA1UdDwEB/wQEAwICpDAPBgNVHRMBAf8EBTADAQH/MB0GA1UdDgQWBBTJ
xfk7RLSg65nniiCdl67BsyYu5DANBgkqhkiG9w0BAQsFAAOCAYEAEJAl0cc+K45q
KFF8bExhWW9NJ4gLsmgBNmosnlVHv4UFWEElpvIfN8Ieoc0dF5V9jy7uWMx/SCI3
Q/ZHW/xJdic0dGXkzBTze3INT43DLrI/3bFjPlfHiKg9+vcdLhLLbF8Hv9hgAubS
UB+qaAl8QUUZ0MbXDXU+C4gfUS9aMb1HFVbgkf3FAQkftDERziKzexkWpsLyJmHS
Zviqje2YA9gYvAQ/t40mYbomvzwdXi11WVO3VdfvRv/iR6YutXO04wZZIJNDYyHM
+AJ+q3fuonRFgk06e4VZybLyGan4KHIcYyr9eJ71Bf72m71lz2yiiZ+AcnI0Nx9v
7PCR5lrQArF4XXBauUnAkC16FGFdAByxHyVKt5/1Es39aZmIbrNcyrHhPznl/ozt
5Z/xwJ+hMWZtYoCB0qtnll3ua1/on4Ho9l6AesJ/7b+BQRarcbdkPJ/sr1UAL39V
m6G3JGIcyHJnols6PoXQMB2Q92hY39b0XhXhQ5DiL/9XfueM96l1
-----END CERTIFICATE-----
```
And mongodb:
```
juju:PRIMARY> db.externalControllers.count()
0
```

Now let's bootstrap a new controller and migrate model-src2 (grafana integrated with influxdb) to it:
```
juju bootstrap localhost ctrl-target
juju switch ctrl-src2
juju migrate model-src2 ctrl-target
```
Let's check that everything has been migrated correctly and stored in dqlite and not in mogodb:
```
juju switch ctrl-target
juju switch model-src2

lxc exec juju-xxx bash

apt install go-dqlite
dqlite -s 127.0.0.1:17666 controller

dqlite> SELECT * FROM external_controller;
b350dd86-cff7-4e42-8979-f8b40d5907e0|ctrl-src1|-----BEGIN CERTIFICATE-----
MIIEEzCCAnugAwIBAgIVALivOnMNMuM2SgZNOFE1dgDwKoZ3MA0GCSqGSIb3DQEB
CwUAMCExDTALBgNVBAoTBEp1anUxEDAOBgNVBAMTB2p1anUtY2EwHhcNMjMwNzI0
MDcyNzU3WhcNMzMwNzI0MDczMjU3WjAhMQ0wCwYDVQQKEwRKdWp1MRAwDgYDVQQD
EwdqdWp1LWNhMIIBojANBgkqhkiG9w0BAQEFAAOCAY8AMIIBigKCAYEA0/uqd0Ac
aINPKp/lhdCKDAfBFKK5HKIt0GmSXBWa0xHwOwKi7a/h3qUhwAG4xJnLGGZc0WUY
mT8ciaJhM9uH41oopkLOvZnI8GU+D+srb1uyvYHIp9v/yICy/K61glUyoiu4ckf/
KeiPPXEZbd3u2Jz9roZPTvPgfes9Otuft3+82Dx6oW8vA4WZb/CGnRZ/tGr9oQ4M
Qqe+TbNM/kxeafyQ2c0yjVG2v3fdaAVT7PE0d8uiWQxVvcQGAyPHr2RPhhqxILJ+
YXTwKo+v46TlJj33GBKUQJfA2oGbtyv9Y7ybkhWfFsWIEmmckZU0Km97N9aGtnqX
NPF1Q85WG3OmO+odiYQ/nDpC/kL9ohndEA1HK7rzOLIk/VynHXEgzMK1Oi5U9aJH
rZ6eecyqKJ5QOkvsBCiA3zzsqRnIr/xAxnsoVTHmlIHaWU/b0hA/l+9jin+1OpbC
H3e14faaDewcsejBP5VdQjTpM+fo8Hni1lnZutxBB2PXt6yNlYt7HqaPAgMBAAGj
QjBAMA4GA1UdDwEB/wQEAwICpDAPBgNVHRMBAf8EBTADAQH/MB0GA1UdDgQWBBTJ
xfk7RLSg65nniiCdl67BsyYu5DANBgkqhkiG9w0BAQsFAAOCAYEAEJAl0cc+K45q
KFF8bExhWW9NJ4gLsmgBNmosnlVHv4UFWEElpvIfN8Ieoc0dF5V9jy7uWMx/SCI3
Q/ZHW/xJdic0dGXkzBTze3INT43DLrI/3bFjPlfHiKg9+vcdLhLLbF8Hv9hgAubS
UB+qaAl8QUUZ0MbXDXU+C4gfUS9aMb1HFVbgkf3FAQkftDERziKzexkWpsLyJmHS
Zviqje2YA9gYvAQ/t40mYbomvzwdXi11WVO3VdfvRv/iR6YutXO04wZZIJNDYyHM
+AJ+q3fuonRFgk06e4VZybLyGan4KHIcYyr9eJ71Bf72m71lz2yiiZ+AcnI0Nx9v
7PCR5lrQArF4XXBauUnAkC16FGFdAByxHyVKt5/1Es39aZmIbrNcyrHhPznl/ozt
5Z/xwJ+hMWZtYoCB0qtnll3ua1/on4Ho9l6AesJ/7b+BQRarcbdkPJ/sr1UAL39V
m6G3JGIcyHJnols6PoXQMB2Q92hY39b0XhXhQ5DiL/9XfueM96l1
-----END CERTIFICATE-----
```
and mongodb:
```
juju:PRIMARY> db.externalControllers.count()
0
```

### CODE REVIEW

File: apiserver/facades/controller/remoterelations/remoterelations.go [hunk 1]
```diff
@@ -429,7 +445,7 @@ func (api *API) UpdateControllersForModels(args params.UpdateControllersForModel
 			CACert:        cInfo.CACert,
 		}
 
-		if err := api.st.UpdateControllerForModel(controller, modelTag.Id()); err != nil {
+		if err := api.ecService.UpdateExternalControllerModel(ctx, modelTag.Id(), controller.ControllerTag.Id()); err != nil {
```
Comment (nvinuesa) [inline apiserver/facades/controller/remoterelations/remoterelations.go:448]: Main change right here

**Review by manadart — COMMENTED** [review]
Code looks alright, but the QA is not to run unit tests. It's
- Bootstrap 3 controllers, src, dst, cns.
- Add a model to src, deploy ./testcharms/charms/dummy-source with an initial token config, and offer it.
- Add a model to cns, deploy ./testcharms/charms/dummy-sink and relate it to the source.
- Wait for the relation to be established, indicated by the token message.
- Migrate the model from src to dst, then check cns.

For me, the logs repeat the following:
```
controller-0: 12:58:50 DEBUG juju.worker.remoterelations cmr remote controller API addresses: [10.246.27.141:17070]
controller-0: 12:58:50 DEBUG juju.worker.remoterelations cmr received redirect; new API addresses: [10.246.27.178:17070]
controller-0: 12:58:50 ERROR juju.worker.remoterelations cmr error in remote application worker for dummy-source: cannot connect to external controller: opening facade to remote model: updating external controller info: FOREIGN KEY constraint failed
controller-0: 12:59:05 DEBUG juju.worker.remoterelations cmr remote controller API addresses: [10.246.27.141:17070]
controller-0: 12:59:05 DEBUG juju.worker.remoterelations cmr received redirect; new API addresses: [10.246.27.178:17070]
controller-0: 12:59:05 ERROR juju.worker.remoterelations cmr error in remote application worker for dummy-source: cannot connect to external controller: opening facade to remote model: updating external controller info: FOREIGN KEY constraint failed
controller-0: 12:59:20 DEBUG juju.worker.remoterelations cmr remote controller API addresses: [10.246.27.141:17070]
controller-0: 12:59:20 DEBUG juju.worker.remoterelations cmr received redirect; new API addresses: [10.246.27.178:17070]
controller-0: 12:59:20 ERROR juju.worker.remoterelations cmr error in remote application worker for dummy-source: cannot connect to external controller: opening facade to remote model: updating external controller info: FOREIGN KEY constraint failed
controller-0: 12:59:35 DEBUG juju.worker.firewaller starting provider endpoint loop for dummy-source:sink dummy-sink:source on dummy-sink
controller-0: 12:59:35 DEBUG juju.worker.remoterelations cmr remote controller API addresses: [10.246.27.141:17070]
controller-0: 12:59:35 DEBUG juju.worker.remoterelations cmr received redirect; new API addresses: [10.246.27.178:17070]
controller-0: 12:59:35 ERROR juju.worker.remoterelations cmr error in remote application worker for dummy-source: cannot connect to external controller: opening facade to remote model: updating external controller info: FOREIGN KEY constraint failed
```

Status has the offer stuck in an error state:
```
Model    Controller  Cloud/Region  Version      SLA          Timestamp
default  cns         lxd/default   4.0-beta1.1  unsupported  13:05:56+02:00

SAAS          Status  Store  URL
dummy-source  error   src    admin/mover.dummy-source

App         Version  Status  Scale  Charm       Channel  Rev  Exposed  Message
dummy-sink           active      1  dummy-sink             0  no       Token is INITIAL

Unit           Workload  Agent  Machine  Public address  Ports  Message
dummy-sink/0*  active    idle   0        10.246.27.122          Token is INITIAL

Machine  State    Address        Inst id        Base          AZ  Message
0        started  10.246.27.122  juju-17134e-0  ubuntu@20.04      Running
```

**Comment by nvinuesa** [comment]
@manadart as discussed offline, in order for the QA scenario you described to work we would also need to have the changes from https://github.com/juju/juju/pull/15928 and https://github.com/juju/juju/pull/15927. I was planning to add that exact scenario as QA for the last PR which will remove the legacy mongo state, should be landing in the next few days.

**Review by manadart — COMMENTED** [review]
One question mark over the actual statement, but the code looks fine.

The QA steps are actually missing the part where the offer is consumed and the relation is established.

When I consume the offer, then relate grafana to influx, the migrate the _offering_ model, it ends up broken.

The SAAS is in an error state according to model-src2, and the migrated model-src1 has 0/0 consumers for the offer.

**Comment by manadart** [comment]
```
machine-0: 15:15:09 DEBUG juju.worker.dependency "external-controller-updater" manifold worker started at 2023-07-24 13:15:09.514463524 +0000 UTC
machine-0: 15:15:09 DEBUG juju.worker.externalcontrollerupdater killing runner 0xc00356a900
machine-0: 15:15:09 INFO juju.worker.externalcontrollerupdater runner is dying
machine-0: 15:15:09 DEBUG juju.worker.dependency "external-controller-updater" manifold worker stopped: watching external controllers: watching external controllers changes: subscribing to namespace "external_controller": catacomb 0xc0007a61c0 is dying
stack trace:
watching external controllers changes: subscribing to namespace "external_controller": catacomb 0xc0007a61c0 is dying
github.com/juju/juju/api/controller/externalcontrollerupdater.(*Client).WatchExternalControllers:44:
github.com/juju/juju/worker/externalcontrollerupdater.(*updaterWorker).loop:113: watching external controllers
machine-0: 15:15:09 ERROR juju.worker.dependency "external-controller-updater" manifold worker returned unexpected error: watching external controllers: watching external controllers changes: subscribing to namespace "external_controller": catacomb 0xc0007a61c0 is dying
```

File: apiserver/facades/controller/remoterelations/remoterelations.go [hunk 2]
```diff
@@ -19,17 +21,30 @@ import (
 	"github.com/juju/juju/state/watcher"
 )
 
+// ECService provides a subset of the external controller domain
+// service methods.
+type ECService interface {
+	// UpdateExternalControllerModel associates the given modelUUID to the
+	// controller identified by controllerUUID.
+	// If the model already exists but it's associated with other controller,
+	// then it's replaced by controllerUUID, if the provided model is already
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply apiserver/facades/controller/remoterelations/remoterelations.go]: this has been removed

File: domain/externalcontroller/service/service.go [hunk 3]
```diff
@@ -35,6 +35,13 @@ type State interface {
 	// The resulting MigrationControllerInfo contains the list of models
 	// for each controller.
 	ControllersForModels(ctx context.Context, modelUUIDs ...string) ([]crossmodel.ControllerInfo, error)
+
+	// UpdateExternalControllerModel associates the given modelUUID to the
+	// controller identified by controllerUUID.
+	// If the model already exists but it's associated with other controller,
+	// then it's replaced by controllerUUID, if the provided model is already
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply domain/externalcontroller/service/service.go]: this has been removed

File: domain/externalcontroller/state/state.go [hunk 4]
```diff
@@ -249,3 +249,31 @@ WHERE  controller_uuid = ?`
...
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	q := `
+INSERT INTO external_model (uuid, controller_uuid)
+VALUES (?, ?)
+  ON CONFLICT(uuid) DO UPDATE SET controller_uuid=excluded.controller_uuid`[1:]
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply domain/externalcontroller/state/state.go]: So, I have removed this method, and when we import an external controller now we simply call `UpdateExternalController` which will add a new ext controller on the db (same as today with mongodb): https://github.com/juju/juju/pull/15940/files#diff-30941483a2bc431f451ab4730d75174a886518663515bcb287b4f0057491159fR446.
I have re-run QA (both ways, migrating offering model and then migrating consuming model) and everything works, SAAS is in correct status and no errors on the logs.

**Review by manadart — APPROVED** [review]
Thanks for seeing this through. I now see the relation re-established, and proven by changing the configuration for `token`, which is seen on the other side.

**Pull request merged**